[cli/move-new] verify if the package name is valid and replace hyphens with underscores

stefan-mysten opened the pull request:
## Description 

This PR adds a check for the `sui move new` command to ensure that the package name is valid (matches this regex: `[A-Za-z][A-Za-z0-9-_]*`), and it replaces hyphens with underscores in the `address` field of the Move.toml file.

## Test Plan 

Added several tests both in `move-cli` and in `sui-move` crates.
```
$ sui move new test_!
Invalid package naming: a valid package name must start with a letter and can contain only letters, hyphens (-), or underscores (_).
```

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [x] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes
Added a check to ensure that pacakge name is valid for the `sui move new` command (must match this regex: `[A-Za-z][A-Za-z0-9-_]*`). Additionally, all hyphens are replaced with underscore in the `address`' field of the Move.toml file.

tnowacki commented on external-crates/move/tools/move-cli/src/base/new.rs:
```diff
@@ -48,6 +52,19 @@ impl New {
         // TODO warn on build config flags
         let Self { name } = self;
+
+        let valid_identifier_re = match Regex::new(r"^[A-Za-z][A-Za-z\_\-]*$") {
```
```suggestion
        let valid_identifier_re = Regex::new(r"^[A-Za-z][A-Za-z\_\-]*$").map_err(|_| anyhow!(...))?;
```

tnowacki commented on external-crates/move/tools/move-cli/src/base/new.rs:
```diff
@@ -48,6 +52,19 @@ impl New {
+            }
+        };
+
+        if !valid_identifier_re.is_match(name.as_str()) {
```
Is name a `String`? I feel like it is more canonical to do 
```suggestion
        if !valid_identifier_re.is_match(&name) {
```

tnowacki commented on external-crates/move/tools/move-cli/src/base/new.rs:
```diff
@@ -48,6 +52,19 @@ impl New {
+        };
+
+        if !valid_identifier_re.is_match(name.as_str()) {
+            return Err(anyhow!("Invalid package naming: a valid package name must start with a letter and can contain only letters, hyphens (-), or underscores (_)."));
+        }
```
anyhow::ensure!

tnowacki commented on external-crates/move/tools/move-cli/src/base/new.rs:
```diff
@@ -58,6 +75,8 @@ impl New {
         };
         create_dir_all(path.join(SourcePackageLayout::Sources.path()))?;
         let mut w = std::fs::File::create(path.join(SourcePackageLayout::Manifest.path()))?;
+        let name = name.to_string();
```
I thought `name` was already a string?

tnowacki commented on external-crates/move/tools/move-cli/src/base/new.rs:
```diff
@@ -76,6 +95,8 @@ version = \"{version}\"
 [addresses]"
         )?;
         for (addr_name, addr_val) in addrs {
+            let addr_name = addr_name.to_string();
+            let addr_name = addr_name.trim().replace('-', "_");
```
Feels a bit strange. Should we just be requiring this and panicking when it does not match?

stefan-mysten replied:
Not sure I understand what you mean by requiring this and panicking. My understanding of the requirement suggested by @damirka was that we don't want any hyphens in the package names and address names.

tnowacki replied:
But these address names (as far as I can tell) are not supplied by the user, only by the tool itself. So I don't think there should be anything to do?

tnowacki replied:
Do we maybe do something with them just on the Sui side?

tzakian replied:
My understanding is this will either need to be done on the Sui side or here, and I personally would have a preference for here since it's a more general Move restriction and not a Sui specific restriction. 

I think we can get into this type of situation if the user does something like `sui move new my-package` then the `addr` being passed in will be `my-package` which will need to be handled here.

tnowacki reviewed (commented):
Overall looks good! Just need some tests :)

damirka reviewed (approved):
Thank you for this!

tnowacki reviewed (changes requested):
I think we we still need to update the regex with digits. And we will need a new CLI test for this before landing. Otherwise looks good!

stefan-mysten commented:
> I think we we still need to update the regex with digits. And we will need a new CLI test for this before landing. Otherwise looks good!

Fixed the regex issue and added tests.

tnowacki commented on external-crates/move/tools/move-cli/tests/build_tests/simple_new/args.txt:
```diff
@@ -3,3 +3,12 @@ new P1
 new P2 -p other_dir
 > cat other_dir/Move.toml
+
+new Move_Package-Test123
+> cat Move_Package-Test123/Move.toml
```
Do we not have a way of binding the address?

tnowacki replied:
Do we generate anything other than a toml on new?

stefan-mysten commented on external-crates/move/tools/move-cli/src/base/new.rs:
```diff
@@ -20,6 +22,8 @@ pub const MOVE_STDLIB_ADDR_VALUE: &str = "0x1";
 /// Create a new Move package with name `name` at `path`. If `path` is not provided the package
 /// will be created in the directory `name`.
+///
+/// By default, this command allows a strict naming scheme based on this regex: [A-Za-z][0-9A-Za-z_]*, and it will replace hyphens (-) with underscore (_) in the package name and address in the Move.toml file.
```
Fix this regex by copy pasting it from the actual source code.

stefan-mysten commented:
Locally, all tests pass. On the CI, many of them fail :| Not sure how to debug this, yet.

tnowacki commented on external-crates/move/crates/move-cli/src/base/new.rs:
```diff
@@ -20,6 +22,8 @@ pub const MOVE_STDLIB_ADDR_VALUE: &str = "0x1";
 /// Create a new Move package with name `name` at `path`. If `path` is not provided the package
 /// will be created in the directory `name`.
+///
+/// By default, this command allows a strict naming scheme based on this regex: [A-Za-z][A-Za-z0-9-_]*, and it will replace hyphens (-) with underscore (_) in the address field in the Move.toml file.
```
This comment is a bit weird for the move-cli side since it does not generate an address

tnowacki replied:
Is there a space to move this to the sui-move side of things?

stefan-mysten replied:
You're right. Let me try to move it and see how it goes!

tnowacki commented on crates/sui-move/src/main.rs:
```diff
@@ -40,10 +40,10 @@ struct Args {
     pub run_bytecode_verifier: bool,
     /// If true, print build diagnostics to stderr--no printing if false
-    #[clap(long = "path", short = 'p', global = true)]
+    #[clap(long = "diagnostics", global = true)]
```
Oh shit

stefan-mysten replied:
Hmm, could you elaborate? 🍡

tnowacki reviewed (approved):
Looks like some of the test failures are for sui move new, so might need to regenerate an exp file? 
Some small comments otherwise.
Thanks for picking this up!
